compiler/rustc_hir_analysis/src/impl_wf_check.rs

```diff
@@ -229,8 +235,99 @@ pub(crate) fn enforce_impl_non_lifetime_params_are_constrained(
                 const_param_note2: const_param_note,
             });
             diag.code(E0207);
+            suggest_to_remove_or_use_generics(tcx, &mut diag, impl_def_id, &param);
             res = Err(diag.emit());
         }
     }
     res
 }
+
+fn suggest_to_remove_or_use_generics(
+    tcx: TyCtxt<'_>,
+    diag: &mut Diag<'_>,
+    impl_def_id: LocalDefId,
+    param: &GenericParamDef,
+) {
+    let node = tcx.hir().get_if_local(impl_def_id.into()).expect("cannot get `Node`");
+    let hir_impl = if let rustc_hir::Node::Item(item) = node {
+        if let rustc_hir::ItemKind::Impl(imp) = item.kind { Some(imp) } else { None }
+    } else {
+        None
+    }
+    .expect("cannot take `Impl` in a impl block");
+
+    let (index, hir_param) = hir_impl
+        .generics
+        .params
+        .iter()
+        .enumerate()
+        .find(|(_, par)| par.name.ident().name == param.name)
+        .unwrap();
+    let mut suggestions = vec![];
+
+    let is_impl_generic = |par: &&GenericParam<'_>| match par.kind {
+        GenericParamKind::Type { .. }
+        | GenericParamKind::Const { .. }
+        | GenericParamKind::Lifetime { kind: LifetimeParamKind::Explicit } => true,
+        _ => false,
+    };
+    // Suggestion for removing the type parameter.
+    suggestions.push(vec![(
+        // Find the span of the type parameter.
+        if let Some(prev) = hir_impl.generics.params[..index].iter().rfind(is_impl_generic) {
+            let mut span = prev.span;
+
+            // Consider the span of the bounds with the generic parameter when there is.
+            if let Some(predicate) = hir_impl.generics.predicates.iter().find(|pred| {
+                if let WherePredicateKind::BoundPredicate(WhereBoundPredicate {
+                    origin: PredicateOrigin::GenericParam,
+                    bounded_ty,
+                    ..
+                }) = pred.kind
+                {
+                    bounded_ty.span == prev.span
+                } else {
+                    false
+                }
+            }) {
+                span = span.to(predicate.span)
+            };
+
+            span.shrink_to_hi().to(hir_param.span)
+        } else if let Some(next) =
+            hir_impl.generics.params[index + 1..].iter().find(is_impl_generic)
+        {
+            hir_param.span.until(next.span)
+        } else {
+            // Remove also angle brackets <> when there is just ONE generic parameter.
+            hir_impl.generics.span
+        },
```

Comment:
Do you mind making all this a method on `Generics`? For instance `fn span_for_param_removal(&self, param: usize) -> Span`.